homeassistant/components/rachio.py

```diff
@@ -0,0 +1,282 @@
+
+    def __init__(self, hass, rachio):
+        """Create an object from the provided API instance."""
+        # Use API token to get user ID
+        self._hass = hass
+        self.rachio = rachio
+
+        response = rachio.person.getInfo()
+        assert int(response[0][KEY_STATUS]) == 200, "API key error"
+        self._id = response[1][KEY_ID]
+
+        # Use user ID to get user data
+        data = rachio.person.get(self._id)
+        assert int(data[0][KEY_STATUS]) == 200, "User ID error"
+        self._username = data[1][KEY_USERNAME]
+        self._controllers = [RachioIro(self._hass, self.rachio, controller)
+                             for controller in data[1][KEY_DEVICES]]
+        _LOGGER.info('Using Rachio API as user "%s"', self._username)
+
+    @property
+    def user_id(self) -> str:
+        """Get the user ID as defined by the Rachio API."""
+        return self._id
+
+    @property
+    def username(self) -> str:
+        """Get the username as defined during account registration."""
+        return self._username
+
+    @property
+    def controllers(self) -> list:
+        """Get a list of controllers managed by this account."""
+        return self._controllers
+
+
+class RachioIro(object):
+    """Represent a Rachio Iro."""
+
+    def __init__(self, hass, rachio, data):
+        """Initialize a Rachio device."""
+        self.hass = hass
+        self.rachio = rachio
+        self._id = data[KEY_ID]
+        self._name = data[KEY_NAME]
+        self._zones = data[KEY_ZONES]
+        self._init_data = data
+        _LOGGER.info('%s has ID "%s"', str(self), self.controller_id)
+
+        # Listen for all updates
+        self._init_webhooks()
+
+    def _init_webhooks(self) -> None:
+        """Start getting updates from the Rachio API."""
+        # Choose which events to listen for and get their IDs
+        event_types = []
+        for event_type in self.rachio.notification.getWebhookEventType()[1]:
+            if event_type[KEY_NAME] in LISTEN_EVENT_TYPES:
+                event_types.append({"id": event_type[KEY_ID]})
+
+        # Register to listen to these events from the device
+        url = self.rachio.webhook_url
+        new_webhook = self.rachio.notification.postWebhook(self.controller_id,
+                                                           WEBHOOK_CONST_ID,
+                                                           url, event_types)
+        # Save ID for deletion at shutdown
+        current_webhook_id = new_webhook[1][KEY_ID]
+
+        def _deinit_webhooks(event) -> None:
+            """Stop getting updates from the Rachio API."""
+            webhooks = self.rachio.notification.getDeviceWebhook(
+                self.controller_id)[1]
+            for webhook in webhooks:
+                if webhook[KEY_EXTERNAL_ID] == WEBHOOK_CONST_ID or\
+                        webhook[KEY_ID] == current_webhook_id:
+                    self.rachio.notification.deleteWebhook(webhook[KEY_ID])
```

Comment:
What happens if this callback does not get called? Is there a way we can query existing webhooks and make sure there is only 1 configured?

Reply:
I have added a line to remove all existing HA webhooks on startup. It should only be 1 extra startup API call if they were removed correctly on shutdown. In my testing, they have always been removed correctly at shutdown (even sending a series of ^C!)